### PR TITLE
builder/amazon: Add suppport for vpc_filter, subnet_filter, and security_group_filter

### PR DESCRIPTION
**Pull request by rickard-von-essen**
Support for discovering VPC's, Subnets, and SG's using filters.

Example:

```json
      "vpc_filter": {
        "filters": {
          "tag:Type": "test"
        }
      },
      "subnet_filter": {
        "filters": {
          "tag:Role": "Test"
        },
        "most_free": true
      },
      "security_group_filter": {
        "filters": {
          "tag:Class": "Packer"
        }
      },
```

Closes: #6339
Closes: #4586

TODO:
  - [x] Docs
  - [x] Implement this for all applicable Amazon builders
  - [x] Add `security_groups_filter`
  - [x] Translate previous VPC and AZ discovery into filters.
  - [x] Remove passing VPC, SG's, and Subnets as parameters, use state bag.
  - [x] Clean up TODO's

### CODE REVIEW

**Comment by rickard-von-essen** [comment]
I'm going away for 10 days, so this will take some time to get the finishing touch.

**Comment by SwampDragons** [comment]
linux build 
[packer.zip](https://github.com/hashicorp/packer/files/2319479/packer.zip)

**Comment by SwampDragons** [comment]
windows build: 
[packer.zip](https://github.com/hashicorp/packer/files/2319486/packer.zip)

**Comment by rickard-von-essen** [comment]
It would be great if someone could try this with _EC2 Classic_.

**Comment by azr** [comment]
hey @rickard-von-essen , I resolved a conflict through github's UI which added a merge commit, you are very free to remove it should it be annoying 🙂 

Will try to understand, review & test a bit now. 🤓

File: builder/amazon/common/step_network_info.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+func (a subnetsSort) Len() int      { return len(a) }
+func (a subnetsSort) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a subnetsSort) Less(i, j int) bool {
+	return *a[i].AvailableIpAddressCount < *a[j].AvailableIpAddressCount
+}
+
+// Returns the most recent AMI out of a slice of images.
+func mostFreeSubnet(subnets []*ec2.Subnet) *ec2.Subnet {
+	sortedSubnets := subnets
+	sort.Sort(subnetsSort(sortedSubnets))
```
Comment (azr) [inline builder/amazon/common/step_network_info.go]: This is not doing exactly what you expect here.
The passed in list will also be sorted

https://play.golang.org/p/wQCtnCBysBP

Comment (azr) [reply builder/amazon/common/step_network_info.go]: but given the usage, I think this is ok 🙂

Comment (rickard-von-essen) [reply builder/amazon/common/step_network_info.go]: Yeah this is a bit over-engineered.

**Comment by tgassman** [comment]
Is there anyway to use a single Tag and Filter based on Values in that tag?  Add SGs that match either value in defined tag.

Like
"tag:AllowedUse": 

Allowed values "Internal", and "Public"

**Pull request merged**